lib/nostrum/struct/channel.ex

```diff
@@ -240,47 +575,107 @@ defmodule Nostrum.Struct.Channel do
-          | channel_category
+          | guild_category_channel
 
   @typedoc """
-  A `Nostrum.Struct.Channel` that represents a text channel
+  All valid text channels.
   """
+  @typedoc deprecated: "See t.0"
   @type text_channel ::
           guild_text_channel
           | dm_channel
           | group_dm_channel
 
   @typedoc """
   A `Nostrum.Struct.Channel` that represents a voice channel
+
   """
+  @typedoc deprecated: "See t.0"
   @type voice_channel :: guild_voice_channel
 
+  @typedoc """
+  All valid channel types.
+  """
   @type t ::
           guild_text_channel
           | dm_channel
           | guild_voice_channel
           | group_dm_channel
-          | channel_category
+          | guild_category_channel
+          | guild_news_channel
+          | guild_store_channel
+          | guild_news_thread_channel
+          | guild_public_thread_channel
+          | guild_private_thread_channel
+          | guild_stage_voice_channel
 
-  @doc ~S"""
-  Formats a `Nostrum.Struct.Channel` into a mention
+  @doc """
+  Convert a channel into a mention.
+
+
+  Handles the conversion of a `Nostrum.Struct.Channel` into the required format to _mention_ the channel within a message.
+
+  ## Parameters
+
+    - channel: `t:Nostrum.Struct.Channel.t/0`
 
   ## Examples
 
-  ```Elixir
-  iex> channel = %Nostrum.Struct.Channel{id: 381889573426429952}
-  ...> Nostrum.Struct.Channel.mention(channel)
-  "<#381889573426429952>"
+  ```elixir
+      > Nostrum.Cache.ChannelCache.get(381889573426429952)
+      > |> Nostrum.Struct.Channel.mention()
+      "<#381889573426429952>"
+
+      > Nostrum.GuildCache.get(81384788765712384)
+      > |> Nostrum.Struct.Channel.mention()
+      "<#Invalid Channel>"
   ```
+
   """
   @spec mention(t) :: String.t()
+  def mention(channel)
   def mention(%__MODULE__{id: id}), do: "<##{id}>"
 
+  def mention(_any), do: "<#Invalid Channel>"
+
+  @doc """
+  Convert a channel into a hyperlink.
+
+  Handles the conversion of a `Nostrum.Struct.Channel` into a [link](#link/1) to that channel.
+
+  > Note: if a user does not already have access to that channel, they will receive "Resource not available" upon following the hyperlink. If you do not know if they have access to the channel or want to grant temporary access it might be safer to use `Nostrum.Api.create_channel_invite/3`
+
+  ## Parameters
+
+  - channel: `t:Nostrum.Struct.Channel.t/0`
+
+  ## Examples
+
+    ```elixir
+        > Nostrum.Cache.ChannelCache.get(381889573426429952)
+        > |> Nostrum.Struct.Channel.link()
+        "https://discordapp.com/channels/#81384788765712384/381889573426429952"
+
+        > Nostrum.GuildCache.get(81384788765712384)
+        > |> Nostrum.Struct.Channel.link()
+        "Channel Not Found"
+    ```
+
+  """
+  @spec link(t) :: String.t()
+  def link(channel)
+
+  def link(%__MODULE__{id: id, guild_id: guild_id}),
+    do: "https://discordapp.com/channels/#{guild_id}/#{id}"
+
+  def link(_any), do: "Channel Not Found"
```

Comment:
Same here as for the comment above: I think we'd want to (or rather: have Elixir) crash explicitly when something else than a channel is passed in, and stick with a single function.

Reply:
I originally wrote it that way and then reconsidered. My reasoning for these was the situation where say, my applications functionality is to post events in a desired channel selected by the user. 

If I want to post a recap of an event, including the channel they were posted in, having it crash or not based on something the user can control (deleting the channel) seemed a bit wrong.

Perhaps the `{:ok, link}`, `{:error, "channel deleted"}` pattern would be better?

Reply:
Added banged version of each to raise a Match Error

Reply:
See my comment above: https://github.com/Kraigie/nostrum/pull/309#discussion_r697898435